elasticsearch/src/main/scala/org/elasticsearch/spark/sql/ElasticsearchXDRelation.scala

```diff
@@ -37,9 +88,256 @@ import org.apache.spark.sql.{Row, SQLContext}
+        var arg = f.productElement(1).toString()
+        if (!strictPushDown) {
+          arg = arg.toLowerCase(Locale.ROOT)
+        }
+        s"""{"query":{"wildcard":{"${f.productElement(0)}":"*$arg"}}}"""
+      }
+
+      case f: Product if isClass(f, "org.apache.spark.sql.sources.StringContains") => {
+        var arg = f.productElement(1).toString()
+        if (!strictPushDown) {
+          arg = arg.toLowerCase(Locale.ROOT)
+        }
+        s"""{"query":{"wildcard":{"${f.productElement(0)}":"*$arg*"}}}"""
+      }
+
+      // the filters below are available only from Spark 1.5.0
+
+      case f: Product if isClass(f, "org.apache.spark.sql.sources.EqualNullSafe") => {
+        var arg = extract(f.productElement(1))
+        if (strictPushDown)
+          s"""{"term":{"${f.productElement(0)}":$arg}}"""
+        else
+          s"""{"query":{"match":{"${f.productElement(0)}":$arg}}}"""
+      }
+
+      case _ => ""
+    }
+  }
+
+    def createDSLFromFilters(filters: Array[Filter], strictPushDown: Boolean) = {
+      filters.map(filter => translateFilter(filter, strictPushDown)).filter(query => StringUtils.hasText(query))
+    }
+
+  }
+
+  // PrunedFilteredScan
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]) = {
+    import CopiedCode._
+    val paramWithScan = mutable.LinkedHashMap[String, String]() ++ parameters
+    paramWithScan += (InternalConfigurationOptions.INTERNAL_ES_TARGET_FIELDS ->
+      StringUtils.concatenate(requiredColumns.asInstanceOf[Array[Object]], StringUtils.DEFAULT_DELIMITER))
+
+    // scroll fields only apply to source fields; handle metadata separately
+    if (cfg.getReadMetadata) {
+      val metadata = cfg.getReadMetadataField
+      // if metadata is not selected, don't ask for it
+      if (!requiredColumns.contains(metadata)) {
+        paramWithScan += (ConfigurationOptions.ES_READ_METADATA -> false.toString())
+      }
+    }
+
+    if (filters != null && filters.size > 0) {
+      if (Utils.isPushDown(cfg)) {
+        if (Utils.LOGGER.isDebugEnabled()) {
+          Utils.LOGGER.debug(s"Pushing down filters ${filters.mkString("[", ",", "]")}")
+        }
+        val filterString = createDSLFromFilters(filters, Utils.isPushDownStrict(cfg))
+
+        if (Utils.LOGGER.isTraceEnabled()) {
+          Utils.LOGGER.trace(s"Transformed filters into DSL ${filterString.mkString("[", ",", "]")}")
+        }
+        paramWithScan += (InternalConfigurationOptions.INTERNAL_ES_QUERY_FILTERS -> IOUtils.serializeToBase64(filterString))
+      }
+      else {
+        if (Utils.LOGGER.isTraceEnabled()) {
+          Utils.LOGGER.trace("Push-down is disabled; ignoring Spark filters...")
+        }
+      }
+    }
+
+    new ScalaXDEsRowRDD(sqlContext.sparkContext, paramWithScan, lazySchema, userSchema)
+  }
+
+  /*override def buildScan(requiredColumns: Array[String], filters: Array[Filter]) = {
```

Comment:
Indeed, to let remind us we should removed the copied code when the ES connector gets updates. The commented code will be valid at the very moment we update its dependency.